### PR TITLE
Helmet HUDs skip over HUDs that are already in effect on the player

### PR DESCRIPTION
**Pull request by Staykeu**
# About the pull request

Makes it so that if you have an active HUD, a helmet visor that does the same thing will not be considered when switching through visors in your helmet. Credit to harryob for his help with the code.

# Explain why it's good for the game

Visors are usually a pain to switch through when their functionality isn't needed, take for example, the squad optic, which does the same job as the headset (HUD-wise). This change makes it so you don't need to switch through unneeded HUDs.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl: Stakeyng
qol: Removes the need to cycle through helmet visors that contain HUD effects that are already active on your character.
/:cl:

### CODE REVIEW

File: code/modules/clothing/head/helmet.dm [hunk 1]
```diff
@@ -680,15 +687,15 @@ GLOBAL_LIST_INIT(allowed_helmet_items, list(
...
-	if(total_visors[1])
-		active_visor = total_visors[1]
+	for(var/obj/item/device/helmet_visor/new_visor in total_visors)
+
+		if(!isnull(GLOB.huds[new_visor.hud_type].hudusers[user]))
+			continue
+
+		active_visor = new_visor
 		toggle_visor(user)
 		return active_visor
```
Comment (harryob) [inline code/modules/clothing/head/helmet.dm:698]:
```suggestion

		return FALSE
```

**Review by harryob — CHANGES_REQUESTED** [review]
for sanity

File: code/modules/clothing/head/helmet.dm [hunk 2]
```diff
@@ -680,15 +687,16 @@ GLOBAL_LIST_INIT(allowed_helmet_items, list(
...
 
-	active_visor = null
-	recalculate_visors(user)
-	return FALSE
-
+		return FALSE
```
Comment (harryob) [inline code/modules/clothing/head/helmet.dm:699]:
```suggestion
	return FALSE
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]
im fine with this but given i my hand in creating the pr we should get someone else to look it over

**Review by Drulikar — CHANGES_REQUESTED** [review]
This is close, but it doesn't fully handle the situation where the next visor is not valid - it skips disabling the current visor. But this isn't your fault; this is actually fairly complex and probably needs to be refactored because its the logic in toggle_visor that's complicating everything. But I believe my suggestions below address the issue I discovered.

For example, equip a helmet w/ medical visor and squad visor. Remove your headset. Toggle visor to squad. Equip your medical glasses. Try to toggle the visor: Icon will change to no visor, but it fails to remove the squad visor and now thinks it cannot toggle to any visors.

However, this requires modification of lines outside of suggestable scope so best I can streamline this is to provide a patch: 
[patch.diff.txt](https://github.com/cmss13-devs/cmss13/files/14130498/patch.diff.txt)

To apply it copy this file to the root of your fork, then open terminal in VSC and enter `git apply patch.diff.txt`

The changes should result in this: 
image

Feel free to ask any questions why I did what I did and to test it.

**Review by Drulikar — APPROVED** [review]

**Pull request merged**